akka-actor/src/main/scala/akka/pattern/AskSupport.scala

```diff
@@ -207,53 +320,153 @@ final class AskableActorSelection(val actorSel: ActorSelection) extends AnyVal {
    * INTERNAL API: for binary compatibility
    */
   protected def ask(message: Any, timeout: Timeout): Future[Any] =
-    internalAsk(message, timeout, ActorRef.noSender)
+    internalSelectionAsk(actorSel, _ ⇒ message, timeout, ActorRef.noSender)
 
   def ask(message: Any)(implicit timeout: Timeout, sender: ActorRef = Actor.noSender): Future[Any] =
-    internalAsk(message, timeout, sender)
+    internalSelectionAsk(actorSel, _ ⇒ message, timeout, sender)
 
   /**
    * INTERNAL API: for binary compatibility
    */
   protected def ?(message: Any)(implicit timeout: Timeout): Future[Any] =
-    internalAsk(message, timeout, ActorRef.noSender)
+    internalSelectionAsk(actorSel, _ ⇒ message, timeout, ActorRef.noSender)
 
   def ?(message: Any)(implicit timeout: Timeout, sender: ActorRef = Actor.noSender): Future[Any] =
-    internalAsk(message, timeout, sender)
+    internalSelectionAsk(actorSel, _ ⇒ message, timeout, sender)
+}
+
+/*
+ * Implementation class of the “ask” pattern with explicit sender enrichment of ActorSelection
+ */
+final class ExplicitlyAskableActorSelection(val actorSel: ActorSelection) extends AnyVal {
+
+  /**
+   * INTERNAL API: for binary compatibility
+   */
+  protected def ask(messageFactory: ActorRef ⇒ Any, timeout: Timeout): Future[Any] =
+    internalSelectionAsk(actorSel, messageFactory, timeout, ActorRef.noSender)
+
+  def ask(messageFactory: ActorRef ⇒ Any)(implicit timeout: Timeout, sender: ActorRef = Actor.noSender): Future[Any] =
+    internalSelectionAsk(actorSel, messageFactory, timeout, sender)
 
   /**
    * INTERNAL API: for binary compatibility
    */
-  private[pattern] def internalAsk(message: Any, timeout: Timeout, sender: ActorRef): Future[Any] = actorSel.anchor match {
+  protected def ?(messageFactory: ActorRef ⇒ Any)(implicit timeout: Timeout): Future[Any] =
+    internalSelectionAsk(actorSel, messageFactory, timeout, ActorRef.noSender)
+
+  def ?(messageFactory: ActorRef ⇒ Any)(implicit timeout: Timeout, sender: ActorRef = Actor.noSender): Future[Any] =
+    internalSelectionAsk(actorSel, messageFactory, timeout, sender)
+}
+
+/**
+ * INTERNAL API: for binary compatibility
+ */
+private[pattern] object InternalAsk {
```

Comment:
the methods in here should be moved into the ExplicitlyAskableActorRef/Selection, respectively